libgraphics/src/directwrite-skia.cpp

```diff
@@ -0,0 +1,433 @@
+	{
+	}
+
+	// IUnknown
+	STDMETHOD_(ULONG, AddRef)();
+	STDMETHOD_(ULONG, Release)();
+	STDMETHOD(QueryInterface)(REFIID, void**);
+
+	// IDWritePixelSnapping
+	STDMETHOD(GetCurrentTransform)(void*, DWRITE_MATRIX*);
+	STDMETHOD(GetPixelsPerDip)(void*, FLOAT*);
+	STDMETHOD(IsPixelSnappingDisabled)(void*, BOOL*);
+
+	// IDWriteTextRenderer
+	STDMETHOD(DrawGlyphRun)(void*, FLOAT, FLOAT, DWRITE_MEASURING_MODE, const DWRITE_GLYPH_RUN*, const DWRITE_GLYPH_RUN_DESCRIPTION*, IUnknown*);
+	STDMETHOD(DrawInlineObject)(void*, FLOAT, FLOAT, IDWriteInlineObject*, BOOL, BOOL, IUnknown*);
+	STDMETHOD(DrawStrikethrough)(void*, FLOAT, FLOAT, const DWRITE_STRIKETHROUGH*, IUnknown*);
+	STDMETHOD(DrawUnderline)(void*, FLOAT, FLOAT, const DWRITE_UNDERLINE*, IUnknown*);
+
+private:
+
+	std::atomic<ULONG> m_refcount = 1;
+	Callback m_callback;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+template <class Callback>
+STDMETHODIMP_(ULONG) MCGDWRenderer<Callback>::AddRef()
+{
+	return ++m_refcount;
+}
+
+template <class Callback>
+STDMETHODIMP_(ULONG) MCGDWRenderer<Callback>::Release()
+{
+	ULONG t_remaining = --m_refcount;
+	if (t_remaining == 0)
+		delete this;
+
+	return t_remaining;
+}
+
+template <class Callback>
+STDMETHODIMP MCGDWRenderer<Callback>::QueryInterface(REFIID riid, void** ppvObject)
+{
+	if (riid == __uuidof(IUnknown) || riid == __uuidof(IDWritePixelSnapping) || riid == __uuidof(IDWriteTextRenderer))
+	{
+		*ppvObject = this;
+		return S_OK;
+	}
+
+	// Not a recognised interface
+	return E_NOINTERFACE;
+}
+
+template <class Callback>
+STDMETHODIMP MCGDWRenderer<Callback>::GetCurrentTransform(void* context, DWRITE_MATRIX* transform)
+{
+	// Get the graphics context
+	MCGContextRef t_context = static_cast<MCGContextRef>(context);
+
+	// Get the transform used for drawing
+	MCGAffineTransform t_transform = MCGContextGetDeviceTransform(t_context);
+
+	// Translate to the DirectWrite matrix format
+	transform->m11 = t_transform.a;
+	transform->m12 = t_transform.b;
+	transform->m21 = t_transform.c;
+	transform->m22 = t_transform.d;
+	transform->dx  = t_transform.tx;
+	transform->dy  = t_transform.ty;
+
+	return S_OK;
+}
+
+template <class Callback>
+STDMETHODIMP MCGDWRenderer<Callback>::GetPixelsPerDip(void* context, FLOAT* pixelsPerDip)
+{
+	// Get the graphics context
+	MCGContextRef t_context = static_cast<MCGContextRef>(context);
+
+	// Get the transform being used for drawing
+	MCGAffineTransform t_transform = MCGContextGetDeviceTransform(t_context);
+	
+	// The scale factor can be derived from the transform
+	*pixelsPerDip = sqrt((t_transform.a*t_transform.a
+						+ t_transform.b*t_transform.b
+						+ t_transform.c*t_transform.c
+						+ t_transform.d*t_transform.d) / 2);
+	
+	return S_OK;
+}
+
+template <class Callback>
+STDMETHODIMP MCGDWRenderer<Callback>::IsPixelSnappingDisabled(void* context, BOOL* isDisabled)
+{
+	// DirectWrite recommends this be enabled unless there is a good reason to disable it
```

Comment:
The comment says 'recommends this be enabled', but isDisabled is being set to TRUE - could you clarify the comment?